crates/ruff_python_ast/ast.toml

```diff
@@ -80,39 +80,367 @@ add_suffix_to_is_methods = true
+optional = true
+
+[[Expr.nodes.ExprLambda.fields]]
+name = "body"
+type = "Expr"
+
+[Expr.nodes.ExprIf]
+rustdoc = "/// See also [IfExp](https://docs.python.org/3/library/ast.html#ast.IfExp)"
+
+[[Expr.nodes.ExprIf.fields]]
+name = "test"
+type = "Expr"
+
+[[Expr.nodes.ExprIf.fields]]
+name = "body"
+type = "Expr"
+
+[[Expr.nodes.ExprIf.fields]]
+name = "orelse"
+type = "Expr"
+
+[Expr.nodes.ExprDict]
+rustdoc = "/// See also [Dict](https://docs.python.org/3/library/ast.html#ast.Dict)"
+
+[[Expr.nodes.ExprDict.fields]]
+name = "items"
+type = "DictItem"
+seq = true
+
+[Expr.nodes.ExprSet]
+rustdoc = "/// See also [Set](https://docs.python.org/3/library/ast.html#ast.Set)"
+
+[[Expr.nodes.ExprSet.fields]]
+name = "elts"
+type = "Expr"
+seq = true
+
+[Expr.nodes.ExprListComp]
+rustdoc = "/// See also [ListComp](https://docs.python.org/3/library/ast.html#ast.ListComp)"
+
+[[Expr.nodes.ExprListComp.fields]]
+name = "elt"
+type = "Expr"
+
+[[Expr.nodes.ExprListComp.fields]]
+name = "generators"
+type = "Comprehension"
+seq = true
+
+[Expr.nodes.ExprSetComp]
+rustdoc = "/// See also [SetComp](https://docs.python.org/3/library/ast.html#ast.SetComp)"
+
+[[Expr.nodes.ExprSetComp.fields]]
+name = "elt"
+type = "Expr"
+
+[[Expr.nodes.ExprSetComp.fields]]
+name = "generators"
+type = "Comprehension"
+seq = true
+
+[Expr.nodes.ExprDictComp]
+rustdoc = "/// See also [DictComp](https://docs.python.org/3/library/ast.html#ast.DictComp)"
+
+[[Expr.nodes.ExprDictComp.fields]]
+name = "key"
+type = "Expr"
+
+[[Expr.nodes.ExprDictComp.fields]]
+name = "value"
+type = "Expr"
+
+[[Expr.nodes.ExprDictComp.fields]]
+name = "generators"
+type = "Comprehension"
+seq = true
+
+[Expr.nodes.ExprGenerator]
+rustdoc = "/// See also [GeneratorExp](https://docs.python.org/3/library/ast.html#ast.GeneratorExp)"
+
+[[Expr.nodes.ExprGenerator.fields]]
+name = "elt"
+type = "Expr"
+
+[[Expr.nodes.ExprGenerator.fields]]
+name = "generators"
+type = "Comprehension"
+seq = true
+
+[[Expr.nodes.ExprGenerator.fields]]
+name = "parenthesized"
+type = "bool"
+
+[Expr.nodes.ExprAwait]
+rustdoc = "/// See also [Await](https://docs.python.org/3/library/ast.html#ast.Await)"
+
+[[Expr.nodes.ExprAwait.fields]]
+name = "value"
+type = "Expr"
+
+[Expr.nodes.ExprYield]
+rustdoc = "/// See also [Yield](https://docs.python.org/3/library/ast.html#ast.Yield)"
+
+[[Expr.nodes.ExprYield.fields]]
+name = "value"
+type = "Expr"
+optional = true
+
+[Expr.nodes.ExprYieldFrom]
+rustdoc = "/// See also [YieldFrom](https://docs.python.org/3/library/ast.html#ast.YieldFrom)"
+
+[[Expr.nodes.ExprYieldFrom.fields]]
+name = "value"
+type = "Expr"
+
+[Expr.nodes.ExprCompare]
+rustdoc = "/// See also [Compare](https://docs.python.org/3/library/ast.html#ast.Compare)"
+
+[[Expr.nodes.ExprCompare.fields]]
+name = "left"
+type = "Expr"
+
+[[Expr.nodes.ExprCompare.fields]]
+name = "ops"
+type = "Box<[crate::CmpOp]>"
+
+[[Expr.nodes.ExprCompare.fields]]
+name = "comparators"
+type = "Box<[Expr]>"
```

Comment:
Having to explicitly use `Box<[Expr]>` here has the downside that it still requires manually updating all types when changing from one representation to another. 

I also find that `type = Expr` + `seq = true` reads somewhat strangely. It also allows for combination that aren't allowed. E.g. is `seq = true` + `optional = true` supported? I wonder if we should do something like:

```
type = "Expr" // single item or escape hatch 
type = { kind: "sequence", element-type: "Expr", slice: true } 
type = { kind: "required", value: "Expr" }
type = { kind: "optiona", value: "Expr" }
```

I'm not very convinced it is better. Maybe you have ideas to improve it further?

Reply:
Yeah the current format has no restrictions and definitely has some problems like You can use type `Vec<Expr>` and also set the sequence. Honestly I was not sure how specific should I make the types in this stage.
I think your suggestion is needed in the end. For example when we want to provide functions to access data without accessing the fields the exact types are useful.

My feeling is that if you are also unsure we can leave this undecided to when it's needed?